scripts/builtin/imputeByKNN.dml

```diff
@@ -0,0 +1,208 @@
+#compute distance between entire "X" (potentially large) and rows with missing values (hopefully small),
+#method 3
+#sample rows, randomly select subset of rows
+#assuming missing value is very large,
+#compute distance between sample of rows of X (control small) and rows with missing values (if not small),
+
+m_imputeByKNN = function(Matrix[Double] data, String method = "default")
+return(Matrix[Double] result)
+{
+  print("KNN-IMPUTATION SCRIPT")
+
+  #Test data, initial data, will be replaced later with variable data
+  first = matrix ("1 3 4 8", rows = 4, cols = 1)
+  second = matrix ("2 4 6 7", rows = 4, cols = 1)
+  third = matrix ("4 NaN 5 NaN", rows = 4, cols = 1)
+  test_matrix = cbind(first,second,third)
+
+  #Add more data for more test cases
+  fifth = matrix("5 6 7", rows = 1, cols = 3)
+  sixth = matrix("3 4 5", rows = 1 , cols = 3)
+  tm2 = rbind(test_matrix, fifth,sixth)
+
+  #Create a mask for placeholder and to check for missing values
+  masked = is.nan(tm2)
+  print(toString(tm2))
+
+  #Find the column containing multiple missing values
+  missing_col = rowIndexMax(colSums(is.nan(tm2)))
+
+  #change method here temporary "default"/"method2" for testing purposes, will be replaced by user input
+  method = "method3"
+
+  #Impute NaN value with temporary mean value of the column
+  filled_matrix = imputeByMean(tm2, matrix(0, cols = ncol(tm2), rows = 1))
+
+  if(method == "default"){
+    #METHOD 1
+    #Calculate the distance using dist method after imputation with mean
+    distance_matrix = dist(filled_matrix)
+
+    #Change 0 value so rowIndexMin will ignore that diagonal value
+    distance_matrix = replace(target = distance_matrix, pattern = 0, replacement = 999)
+
+    #Get the minimum distance row-wise computation
+    minimum_index = rowIndexMin(distance_matrix)
+
+    #Position of missing values in per row in which column
+    position = rowSums(is.nan(tm2))
+    position = position * minimum_index
+
+    #Filter the 0 out
+    I = (rowSums(is.nan(tm2))!=0)
+    missing = removeEmpty(target=position, margin="rows", select=I)
+
+    #Convert the value indices into 0/1 matrix to find location
+    indices = table(missing, seq(1,nrow(filled_matrix)),odim1=nrow(filled_matrix),odim2=nrow(missing))
+
+    #Replace the index with value
+    imputedValue = t(indices) %*% filled_matrix[,as.scalar(missing_col)]
+
+    #Get the index location of the missing value
+    x = rowSums(is.nan(tm2))
+    missing_indices = seq(1, nrow(x)) * x
+
+    #Put the replacement value in the missing indices
+    I2 = removeEmpty(target=missing_indices, margin="rows")
+    R = table(I2,1,imputedValue,odim1 = nrow(tm2), odim2=1)
+
+    #Replace the masked column with to be imputed Value
+    masked[,as.scalar(missing_col)] = masked[,as.scalar(missing_col)] * R
+
+    #Impute the value
+    result = replace(target = tm2, pattern = NaN, replacement = 0)
+    result = result + masked
+    print("Result method1")
+    print(toString(result))
```

Comment:
move common code outside the method-specific branches (at the end of the builtin function).